plugins/inputs/suricata/suricata.go

```diff
@@ -0,0 +1,229 @@
+package suricata
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const (
+	// InBufSize is the input buffer size for JSON received via socket.
+	// Set to 10MB, as depending on the number of threads the output might be
+	// large.
+	InBufSize = 10 * 1024 * 1024
+)
+
+// Suricata is a Telegraf input plugin for Suricata runtime statistics.
+type Suricata struct {
+	sync.Mutex
+
+	Source    string `toml:"source"`
+	Delimiter string `toml:"delimiter"`
+
+	InputListener *net.UnixListener
+	Ctx           context.Context
+	Cancel        context.CancelFunc
+
+	wg sync.WaitGroup
+}
+
+// Description returns the plugin description.
+func (s *Suricata) Description() string {
+	return "Suricata stats plugin"
+}
+
+const sampleConfig = `
+  ## Data sink for Suricata stats log
+  # This is expected to be a filename of a
+  # unix socket to be created for listening.
+  # Will be overwritten if a socket or file
+  # with that name already exists.
+  source = "/var/run/suricata-stats.sock"
+  # Delimiter for flattening field keys, e.g. subitem "alert" of "detect"
+  # becomes "detect_alert" when delimiter is "_".
+  delimiter = "_"
+`
+
+// SampleConfig returns a sample TOML section to illustrate configuration
+// options.
+func (s *Suricata) SampleConfig() string {
+	return sampleConfig
+}
+
+// Start initiates background collection of JSON data from the socket
+// provided to Suricata.
+func (s *Suricata) Start(acc telegraf.Accumulator) error {
+	var err error
+	s.Lock()
+	defer s.Unlock()
+	if s.InputListener == nil {
+		s.InputListener, err = net.ListenUnix("unix", &net.UnixAddr{
+			Name: s.Source,
+			Net:  "unix",
+		})
+		if err != nil {
+			return err
+		}
+		s.Ctx, s.Cancel = context.WithCancel(context.Background())
+		s.InputListener.SetUnlinkOnClose(true)
+		s.wg.Add(1)
+		go s.handleServerConnection(acc)
+	}
+	return nil
+}
+
+// Stop causes the plugin to cease collecting JSON data from the socket provided
+// to Suricata.
+func (s *Suricata) Stop() {
+	s.Lock()
+	defer s.Unlock()
+	s.InputListener.Close()
+	if s.Cancel != nil {
+		s.Cancel()
+	}
+	s.wg.Wait()
+	s.InputListener.Close()
+	s.InputListener = nil
+}
+
+func (s *Suricata) readInput(acc telegraf.Accumulator, conn net.Conn) {
+	reader := bufio.NewReaderSize(conn, InBufSize)
+	for {
+		select {
+		case <-s.Ctx.Done():
+			return
+		default:
+			line, rerr := reader.ReadBytes('\n')
+			if rerr == nil {
+				s.parse(acc, line)
+			} else if rerr == io.EOF {
+				return
+			}
+		}
+	}
+}
+
+func (s *Suricata) handleServerConnection(acc telegraf.Accumulator) {
+	var err error
+	defer s.wg.Done()
```

Comment:
Just a suggestion, but I usually prefer to locate all the goroutine functionality together outside the goroutine.  This would allows the function to be called not as a goroutine.  I also pass in the Context and don't save it around on the struct, keeping only the CancelFunc.  At the call point I do:
```
s.wg.Add(1)
go func() {
    defer s.wg.Done()
    handleServerConnection(ctx, acc)
}
```

Reply:
Done.